File: src/main/kotlin/net/namibsun/nuztrack/transfer/RunsTransfer.kt
package net.namibsun.nuztrack.transfer

import net.namibsun.nuztrack.constants.ValidationException
import net.namibsun.nuztrack.constants.enums.ErrorMessages
import net.namibsun.nuztrack.constants.enums.Games
import net.namibsun.nuztrack.constants.enums.Rules
import net.namibsun.nuztrack.data.NuzlockeRun

data class NuzlockeRunTO(
    val id: Long,
    val userName: String,
    val name: String,
    val game: GameTO,
    val rules: List<String>,
    val customRules: List<String>,
    val status: String
) {
    companion object {
        fun fromNuzlockeRun(run: NuzlockeRun): NuzlockeRunTO {
            return NuzlockeRunTO(
                id = run.id,
                userName = run.userName,
                name = run.name,
                game = GameTO.fromGame(run.game),
                rules = run.rules.map { it.name.uppercase() },
                customRules = run.customRules,
                status = run.status.name
            )
        }
    }
}

data class CreateNuzlockeRunTO(
    val name: String,
    val game: String,
    val rules: List<String>,
    val customRules: List<String>
) {
    fun validate() {
        if (name == "") {
            throw ValidationException(ErrorMessages.EMPTY_NAME)
        }
        Games.valueOfWithChecks(game)
        rules.map { Rules.valueOfWithChecks(it) }
    }
}

File: frontend/src/data/runs/runs.transfer.ts
import {GameTO} from "../games/games.transfer";

export interface NuzlockeRunTO {
    id: number,
    username: string,
    name: string,
    game: GameTO,
    rules: string[],
    customRules: string[],
    status: string
}

export interface NuzlockeRunCreatorTO {
    name: string,
    game: string,
    rules: string[],
    customRules: string[]
}

export interface CreateMultiRunTO {
    runId: number,
    options: string[],
    game: string,
    name: string
}

export interface MultiRunOptionTO {
    key: string
    description: string
}